### ISSUE TITLE
WARNING: Class 'x.y.z' could not be resolved

### ISSUE THREAD

**Issue by tmohme** [jlink/jqwik#27]
## Test Execution Problem
Under certain circumstances we get the following warning 
> net.jqwik.discovery.HierarchicalJavaResolver resolveClass
WARNUNG: Class 'x.y.z' could not be resolved

without understanding why we get it. All tests seemingly were executed successfully.
This happens only with a very specific configuration of maven-surefire, implicating that the root cause probably lies on their ground.  
I created a demo project with which the warning can be triggered: https://github.com/tmohme/jqwik-warning
For reproduction please checkout the project and build it with 'mvn clean test "-Dtest=org.example.*.java"'.
Any change to the class-loader related configuration of the maven-surefire-plugin lets the warning disappear. Also there's no warning, when you simply execute 'mvn clean test'. 
Maybe this issue should be "forwarded" to the maven-surefire-plugin.

## Desired behaviour
In the best case, we shouldn't get the warning.

## Discussion
If the root cause lies in wrong/weird class loading behaviour of surefire, the warning should be accompanied by a hint what the possible consequences (for the test execution) are.
Maybe documenting this restriction would help.

**Comment by jlink** [comment]
## TLDR

It's just a warning that the classes `IdentifierEqualsContract` and `IdentifierLiteralContract` are being tried as test containers but they are not. The warning can be ignored.

## Longer Explanation

The maven option "-Dtest=org.example.*.java" instructs maven to consider all types (classes and interfaces) that match the pattern as potential test containers. That's why maven sends a "resolve as test class" request to jqwik (or any other registered test engine). Maven is behaving according to contract in this case.

Since those contract classes are interfaces they are not considered to be valid test containers by jqwik _on their own_. They are however considered by jqwik as supertypes of the *Properties classes.

It should be possible to change the "-Dtest=..." pattern in a way to only include classes that end with "Properties" (my adhoc attempty to do that sadly failed though).

## Possible Changes in jqwik

What would you consider a better warning message to make the explanation above more obvious?

**Comment by tmohme** [comment]
Thank you for the explanation.  
For me, what you wrote in the "TLDR" section of your comment would be sufficient.
Something like `jqwik was instructed to try class 'x.y.z' as a test container, but it is not (a test container on its own)`.  
The current message `Class 'x.y.z' could not be resolved` creates the fear that the tests might not be executed as expected and thus leave you uncovered.  
Also the severity could be changed from `warning` to `info` if the message can (always) safely be ignored (= there's no need to act upon it).  
Personally I regard `warning` message as important enough to investigate their meaning/impact, while `info` messages can safely be ignored.

**Comment by stse-zuehlke-com** [comment]
Also from me thank you for your explanation.

I would like to suggest that nothing is logged at all in such cases, because nobody would expect that such interfaces/classes are treated as tests or is the message logged by maven or surefire?

**Comment by stse-zuehlke-com** [comment]
I think the question is: Should the classes which contain only a contract and not an executable test be treated as test containers?. I have forked @tmohme's example (see https://github.com/stse-zuehlke-com/jqwik-warning) and migrated it to plain junit 5 in order to check how jupiter handles such contract interfaces in this case. When I execute "mvn clean test -Dtest=org.example.*.java" then jupiter handles these interfaces as non test containers without logging a warning or an info.

**Comment by jlink** [comment]
Did you enable logging for log level DEBUG when you called maven?

**Comment by stse-zuehlke-com** [comment]
Good question. Nope. But also if I execute "mvn clean test -Dtest=org.example.*.java" with "-X" there are not any debug messages regarding these classes.

**Comment by jlink** [comment]
Jupiter is doing logging through some additional redirection for reasons not obvious to me.

Anyway, what I will do is change the log level to INFO and adapt the message to something hopefully more understandable like:

`request to resolve class 'x.y.z' as a test container, but it is not a container or does not contain tests`

**Comment by jlink** [comment]
Hopefully resolved in https://github.com/jlink/jqwik/commit/87d31846eccd11c7b5535d78496fffde6565068e

**Comment by jlink** [comment]
You might want to try it in 0.8.13-SNAPSHOT

**Comment by tmohme** [comment]
Confirm that 0.8.13 produces a more helpful message (on INFO level).
